modules/gdscript/gdscript_compiler.cpp

```diff
@@ -939,19 +951,38 @@ GDScriptCodeGenerator::Address GDScriptCompiler::_parse_expression(CodeGen &code
 							// Check for a property.
 							if (n->base->type == GDScriptParser::Node::IDENTIFIER) {
 								GDScriptParser::IdentifierNode *identifier = static_cast<GDScriptParser::IdentifierNode *>(n->base);
-								StringName var_name = identifier->name;
+								var_name = identifier->name;
 								if (_is_class_member_property(codegen, var_name)) {
 									assign_class_member_property = var_name;
-								} else if (!_is_local_or_parameter(codegen, var_name) && (codegen.script->member_indices.has(var_name) || codegen.script->static_variables_indices.has(var_name))) {
-									is_member_property = true;
-									is_static = codegen.script->static_variables_indices.has(var_name);
-									const GDScript::MemberInfo &minfo = is_static ? codegen.script->static_variables_indices[var_name] : codegen.script->member_indices[var_name];
-									member_property_setter_function = minfo.setter;
-									member_property_has_setter = member_property_setter_function != StringName();
-									member_property_is_in_setter = member_property_has_setter && member_property_setter_function == codegen.function_name;
-									target_member_property.mode = is_static ? GDScriptCodeGenerator::Address::STATIC_VARIABLE : GDScriptCodeGenerator::Address::MEMBER;
-									target_member_property.address = minfo.index;
-									target_member_property.type = minfo.data_type;
+								} else if (!_is_local_or_parameter(codegen, var_name)) {
+									if (codegen.script->member_indices.has(var_name)) {
+										is_member_property = true;
+										is_static = false;
+										const GDScript::MemberInfo &minfo = codegen.script->member_indices[var_name];
+										member_property_setter_function = minfo.setter;
+										member_property_has_setter = member_property_setter_function != StringName();
+										member_property_is_in_setter = member_property_has_setter && member_property_setter_function == codegen.function_name;
+										target_member_property.mode = GDScriptCodeGenerator::Address::MEMBER;
+										target_member_property.address = minfo.index;
+										target_member_property.type = minfo.data_type;
+									} else {
+										// Try static variables.
+										GDScript *scr = codegen.script;
+										while (scr) {
+											if (scr->static_variables_indices.has(var_name)) {
+												is_member_property = true;
+												is_static = true;
+												const GDScript::MemberInfo &minfo = scr->static_variables_indices[var_name];
+												member_property_setter_function = minfo.setter;
+												member_property_has_setter = member_property_setter_function != StringName();
+												member_property_is_in_setter = member_property_has_setter && member_property_setter_function == codegen.function_name;
+												static_var_class = codegen.add_constant(scr);
```

Comment:
This might be problematic if `scr` refers to the script being compiled, because it will make a reference to itself so it will never be freed. In such case it should use a `CLASS` address instead to use the current script.

There are other instances of this pattern in the compiler, all of them have the same issue.

Reply:
The problem is that the method can be inherited (see #77331) in which case `CLASS` will point to the current class, not the one the static variable is located. This will result in an error (the child class variable with the same index will be used or the `GD_ERR_BREAK` index check will be triggered).

So the `CLASS` address can only be used in the static initializer, since in this case the current class always matches.

Also, if these constants are counted, then we can still have problems in case of cyclic references. But I think they shouldn't count because it's `*GDScript` (not `Ref<GDScript>`)?

Reply:
I tested it, it does add an extra reference and doesn't depend on `GDScript*` vs `Ref<GDScript>` (some `Variant`/`Object` magic?). But we could store `ObjectID` instead of `GDScript*` (this is how `WeakRef` is implemented).

I also found that we **already** have an extra reference to each script somewhere, the bug is reproduced in master and in 4.0.3. Later I will check/open an issue.

Reply:
> I tested it, it does add an extra reference and doesn't depend on `GDScript*` vs `Ref<GDScript>` (some `Variant`/`Object` magic?)

All constants are Variants, if you put a RefCounted object it will increment the count (it does not matter if the `Ref<>` wrapper is used, the Variant itself act as the wrapper). 

> I also found that we **already** have an extra reference to each script somewhere, the bug is reproduced in master and in 4.0.3. Later I will check/open an issue.

I mentioned this in the static variables PR. But I don't think there's an open issue yet. There's a bit of weird stuff happening in the GDScriptCache so the issue is a bit complicated.

Reply:
> The problem is that the method can be inherited (see #77331) in which case `CLASS` will point to the current class, not the one the static variable is located

What I meant is to use `CLASS` only in the case where it is in current class. If it is from a super class then it's fine because there's already a reference to it anyway.

Reply:
> I mentioned this in the static variables PR. But I don't think there's an open issue yet.

* See #77513.